native/core/src/execution/shuffle/shuffle_writer.rs

```diff
@@ -667,175 +740,322 @@ impl Debug for ShuffleRepartitioner {
-                return Ok(AppendRowStatus::StartIndex(start));
+    /// Consumes buffered batches and return a concatenated batch if successful
+    fn concat_buffered_batches(&mut self) -> Result<Option<RecordBatch>> {
+        if self.buffered_batches.is_empty() {
+            Ok(None)
+        } else if self.buffered_batches.len() == 1 {
+            let batch = self.buffered_batches.remove(0);
+            self.num_buffered_rows = 0;
+            Ok(Some(batch))
+        } else {
+            let schema = &self.buffered_batches[0].schema();
+            match arrow::compute::concat_batches(schema, self.buffered_batches.iter()) {
+                Ok(concatenated) => {
+                    self.buffered_batches.clear();
+                    self.num_buffered_rows = 0;
+                    Ok(Some(concatenated))
+                }
+                Err(e) => Err(DataFusionError::ArrowError(
+                    e,
+                    Some(DataFusionError::get_back_trace()),
+                )),
             }
+        }
+    }
+}
 
-            let mut repart_timer = metrics.repart_time.timer();
-            self.active
-                .iter_mut()
-                .zip(columns)
-                .for_each(|(builder, column)| {
-                    append_columns(builder, column, &indices[start..end], column.data_type());
-                });
-            self.num_active_rows += end - start;
-            repart_timer.stop();
-            start = end;
+#[async_trait::async_trait]
+impl ShufflePartitioner for SinglePartitionShufflePartitioner {
+    async fn insert_batch(&mut self, batch: RecordBatch) -> Result<()> {
+        let start_time = Instant::now();
+        let num_rows = batch.num_rows();
+
+        if num_rows > 0 {
+            self.metrics.data_size.add(batch.get_array_memory_size());
+            self.metrics.baseline.record_output(num_rows);
+
+            if num_rows >= self.batch_size || num_rows + self.num_buffered_rows > self.batch_size {
+                let concatenated_batch = self.concat_buffered_batches()?;
+
+                let write_start_time = Instant::now();
 
-            if self.num_active_rows >= self.batch_size {
-                self.flush(metrics)?;
+                // Write the concatenated buffered batch
+                if let Some(batch) = concatenated_batch {
+                    self.output_data_writer
+                        .write(&batch, &self.metrics.encode_time)?;
+                }
+
+                if num_rows >= self.batch_size {
+                    // Write the new batch
+                    self.output_data_writer
+                        .write(&batch, &self.metrics.encode_time)?;
+                } else {
+                    // Add the new batch to the buffer
+                    self.add_buffered_batch(batch);
+                }
+
+                self.metrics
+                    .write_time
+                    .add_duration(write_start_time.elapsed());
```

Comment:
The native `write_time` metric maps to Spark's `shuffleWriteTime` and is intended to be the amount of time writing to disk, excluding any cost of encoding. This is one example where the metric changes in this PR to include the cost of the call to `self.output_data_writer.write` which performs encoding (and may just write to a buffer and not to disk). I'd suggest passing the `write_time` metric into `self.output_data_writer.write` along with the `encode_time` metric so that we only increment the write time for the the write-to-disk operation.

Reply:
Added write_time parameter to the write method of `BufBatchWriter`.